### PR TITLE
kvclient: handle RangeKeyMismatchError correctly

### LINKED ISSUE #105294: distsender: do not evict range descriptor on stale RangeKeyMismatchError
**Describe the problem**

When the `DistSender` receives a `RangeKeyMismatchErrors` it simply terminates its inner loop (`sendToReplicas`) and evicts the range descriptor. In doing so, it updates its range cache with the range descriptors in the error; see:

https://github.com/cockroachdb/cockroach/blob/5dd802945591fce48b1e357e4a9f91ec13f82057/pkg/kv/kvclient/kvcoord/dist_sender.go#L1895-L1908

Now, if the error is returned by a replica that's lagging (say it hasn't applied a split or a merge yet), we'll end up needlessly evicting the range descriptor. Worse yet, if on the subsequent retry the sender picks this replica[1] as the first one to route to again, we'll find ourselves in an infinite loop until this replica is caught up. 

**Proposed solution**

We know when a `RangeKeyMismatchError` is originating from a replica that isn't caught up yet, by virtue of the descriptor generation in the first entry of its `Ranges` slice. Currently, `sendToReplicas` short circuits without peeking into this slice:

https://github.com/cockroachdb/cockroach/blob/5dd802945591fce48b1e357e4a9f91ec13f82057/pkg/kv/kvclient/kvcoord/dist_sender.go#L2533-L2536

Maybe it should. That is, it should only terminate its iteration if the `RangeKeyMismatch` error is coming from a replica that has a descriptor generation newer than what's in the range cache. 

cc @kvoli @andrewbaptist 

----

[1] This can happen if the sender believes this replica has the lease or if it has no lease information and the replica is closest to it.

### PR DESCRIPTION
**Pull request by andrewbaptist**
Previously if a client returned a RangeKeyMismatchError with a lower descriptor generation, the server would remove the previous information from its cache and reload from meta1/2. If this resulted in it choosing the same store to look at first again, it would cycle between reading the up-to-date information from meta1, trying the first store which had a stale range, until the store updated its replica to see the split. This is bad in general but with replica pausing, this can be a long-lived state. This commit changes the behavior to verify the generation has been updated before using the data from the server. p

Epic: none
Fixes: #105294

Release note: None

### CODE REVIEW

**Review by arulajmani — COMMENTED** [review]
Sorry for the (extremely) late review on this one.

> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-:-NnuVm4o3toOAEURpUSK:bph111p)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist)

___
*[`-- commits` line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-NnuOb622Q8PXRB2dP0N):*
nit: reload from meta2 (meta1 points to meta2, meta2 has range descriptors). 
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2414 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-NnuRHyjDzHFXTcLmFG5:-NnuRHyjDzHFXTcLmFG6:b-4j8jvh) ([raw file](https://github.com/cockroachdb/cockroach/blob/8607ad4c151bd720f808e5b5cb63ae0b82d8441e/pkg/kv/kvclient/kvcoord/dist_sender.go#L2414)):*
> ```Go
> 			case *kvpb.RangeKeyMismatchError:
> 				// If the error has strictly more up-to-date information than what we
> 				// currently know about this range, then stop looping so we update our
> ```

nit: consider inverting the condition here to "is the RangeKeyMismatchError coming from a lagging replica; if so, take no action and continue looping". I think that's a more natural way to reason about this, though feel free to disregard if you disagree.
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 590 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-NnuR3pFDGxsL5_hDqb_:-NnuR3pFDGxsL5_hDqba:b-1s6s9m) ([raw file](https://github.com/cockroachdb/cockroach/blob/8607ad4c151bd720f808e5b5cb63ae0b82d8441e/pkg/kv/kvclient/rangecache/range_cache.go#L590)):*
> ```Go
> 	// We have new descriptors and at least one of them overlaps with and is newer
> 	// than our descriptor.
> 	if len(newDescs) > 0 && et.compareDescriptorsLocked(newDescs...) < 0 {
> ```

I don't think we need this -- this case is already handled further down in `insertLocked` when we call `clearOlderOverlappingLocked`. 
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 624 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-NnuPyGwD83JyG_imtBd:-NnuPyGwD83JyG_imtBe:b8rxxhu) ([raw file](https://github.com/cockroachdb/cockroach/blob/8607ad4c151bd720f808e5b5cb63ae0b82d8441e/pkg/kv/kvclient/rangecache/range_cache.go#L624)):*
> ```Go
> 	comparison := -1
> 	for _, r := range newDescs {
> 		if _, err := et.Desc().RSpan().Intersect(r.Desc.RSpan()); err == nil {
> ```

Do we ever expect this function to be called with non-intersecting descriptors? 

If not, let's codify that into an assertion. Then, we can simplify the return value of this function to return a boolean (as -1 won't be possible) .
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 629 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-NnuQYp9-PrRFF6zEETk:-NnuQYp9-PrRFF6zEETl:bmcc3xt) ([raw file](https://github.com/cockroachdb/cockroach/blob/8607ad4c151bd720f808e5b5cb63ae0b82d8441e/pkg/kv/kvclient/rangecache/range_cache.go#L629)):*
> ```Go
> 			// use the latest info we get.
> 			if r.Desc.Generation == et.Desc().Generation {
> 				// We found at least one key where the generation increased.
> ```

nit: Should this comment be in the branch below? 
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 650 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-NnuQaVi5llxdEgg-uzr:-NnuQaVi5llxdEgg-uzs:by82ub3) ([raw file](https://github.com/cockroachdb/cockroach/blob/8607ad4c151bd720f808e5b5cb63ae0b82d8441e/pkg/kv/kvclient/rangecache/range_cache.go#L650)):*
> ```Go
> // descriptor and retry the request while if it is older we assume the server
> // has stale information.
> func (et *EvictionToken) CompareDescriptors(newDescs ...roachpb.RangeInfo) int {
> ```

nit: With my suggestion above to simplify the return value of this function, let's also rename it to be something more specific than `CompareDescriptors`. Maybe, `AreOverlappingDescriptorsNewer` or something similar?
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_test.go` line 5570 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-NnuTcQH2mLrqPzVw_pJ:-NnuTcQH2mLrqPzVw_pK:b-mhtbbo) ([raw file](https://github.com/cockroachdb/cockroach/blob/8607ad4c151bd720f808e5b5cb63ae0b82d8441e/pkg/kv/kvclient/kvcoord/dist_sender_test.go#L5570)):*
> ```Go
> 					br.Error = kvpb.NewError(&kvpb.RangeNotFoundError{})
> 				case 2:
> 					// If the last store is sending success, then send the range key
> ```

This `lastStoreSuccess` bit in the test seems needlessly complicated -- can we simplify the test by always returning success from the last store? We would then only need to test for two cases -- 

1. If the `RangeKeyMismatchError` has an older generation than the client. 
2. If the `RangeKeyMismatchError` has a newer generation than the client. 

In the first case, we'd want to assert that we never exited the `sendToReplicas` loop. So count == 1. In the second, we'd want the ensure we exited and retried, so count == 2.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_test.go` line 5625 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-NnuUKNfEzMGBPIAeTDT:-NnuUKNfEzMGBPIAeTDU:b-tp7wds) ([raw file](https://github.com/cockroachdb/cockroach/blob/8607ad4c151bd720f808e5b5cb63ae0b82d8441e/pkg/kv/kvclient/kvcoord/dist_sender_test.go#L5625)):*
> ```Go
> 			ds := NewDistSender(cfg)
> 			ba := &kvpb.BatchRequest{}
> 			get := &kvpb.GetRequest{}
> ```

nit: you can use `kvpb.NewGet()` here instead.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_test.go` line 5640 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-NnuVh4y7dNNH5xSvKBt:-NnuVh4y7dNNH5xSvKBu:brke2yu) ([raw file](https://github.com/cockroachdb/cockroach/blob/8607ad4c151bd720f808e5b5cb63ae0b82d8441e/pkg/kv/kvclient/kvcoord/dist_sender_test.go#L5640)):*
> ```Go
> 
> 			{
> 				_, err := ds.Send(ctx, ba)
> ```

I think this test can be simplified to only use a single Send.

**Review by andrewbaptist — COMMENTED** [review]
No problem! Glad you got a chance to get to it. I'm making a couple of the comments, but I want to rebase this on top of #117632 which touches some of the same code. I think the fix will be simpler with the refactoring in that PR.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-:-NoNpApaAvtf1GVJOxq-:beztinl)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`-- commits` line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-NnuOb622Q8PXRB2dP0N):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: reload from meta2 (meta1 points to meta2, meta2 has range descriptors).
</blockquote></details>

Thanks! I didn't fully understand this breakdown when I wrote this, but do now.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2414 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-NnuRHyjDzHFXTcLmFG5:-NoOF4TS6VDpv1E-y-5P:bolbh6b) ([raw file](https://github.com/cockroachdb/cockroach/blob/8607ad4c151bd720f808e5b5cb63ae0b82d8441e/pkg/kv/kvclient/kvcoord/dist_sender.go#L2414)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: consider inverting the condition here to "is the RangeKeyMismatchError coming from a lagging replica; if so, take no action and continue looping". I think that's a more natural way to reason about this, though feel free to disregard if you disagree.
</blockquote></details>

Makes sense I inverted the condition. This was easier to do after rebasing on the other branch.
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 590 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-NnuR3pFDGxsL5_hDqb_:-NoOGupU8jTEL22NdPNx:bz9dpwr) ([raw file](https://github.com/cockroachdb/cockroach/blob/8607ad4c151bd720f808e5b5cb63ae0b82d8441e/pkg/kv/kvclient/rangecache/range_cache.go#L590)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I don't think we need this -- this case is already handled further down in `insertLocked` when we call `clearOlderOverlappingLocked`.
</blockquote></details>

It doesn't work without this check. This is because the check below is after we have already called `evictDescLocked` so it is too late.
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 624 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-NnuPyGwD83JyG_imtBd:-NoO-XGp8BeEKNoLFD35:b-nf3ow9) ([raw file](https://github.com/cockroachdb/cockroach/blob/8607ad4c151bd720f808e5b5cb63ae0b82d8441e/pkg/kv/kvclient/rangecache/range_cache.go#L624)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Do we ever expect this function to be called with non-intersecting descriptors? 

If not, let's codify that into an assertion. Then, we can simplify the return value of this function to return a boolean (as -1 won't be possible) .
</blockquote></details>

I'm not positive if this assertion is true or not. I changed to true/false as we can treat non-overlapping as false.
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 629 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-NnuQYp9-PrRFF6zEETk:-NoNza3_8GP1rzBYbCNR:b-d97ndf) ([raw file](https://github.com/cockroachdb/cockroach/blob/8607ad4c151bd720f808e5b5cb63ae0b82d8441e/pkg/kv/kvclient/rangecache/range_cache.go#L629)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: Should this comment be in the branch below?
</blockquote></details>

Moved comment
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 650 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-NnuQaVi5llxdEgg-uzr:-NoNyVLe0OqEezFr0KaJ:b-y4kac) ([raw file](https://github.com/cockroachdb/cockroach/blob/8607ad4c151bd720f808e5b5cb63ae0b82d8441e/pkg/kv/kvclient/rangecache/range_cache.go#L650)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: With my suggestion above to simplify the return value of this function, let's also rename it to be something more specific than `CompareDescriptors`. Maybe, `AreOverlappingDescriptorsNewer` or something similar?
</blockquote></details>

Renamed as suggested, that name is better.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_test.go` line 5570 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-NnuTcQH2mLrqPzVw_pJ:-NoOGqb8-FZkiIHtwEFG:be3iivx) ([raw file](https://github.com/cockroachdb/cockroach/blob/8607ad4c151bd720f808e5b5cb63ae0b82d8441e/pkg/kv/kvclient/kvcoord/dist_sender_test.go#L5570)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

This `lastStoreSuccess` bit in the test seems needlessly complicated -- can we simplify the test by always returning success from the last store? We would then only need to test for two cases -- 

1. If the `RangeKeyMismatchError` has an older generation than the client. 
2. If the `RangeKeyMismatchError` has a newer generation than the client. 

In the first case, we'd want to assert that we never exited the `sendToReplicas` loop. So count == 1. In the second, we'd want the ensure we exited and retried, so count == 2.
</blockquote></details>

Yeah - this is a little complex. I'm going to try and simplify this a little bit.... 
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_test.go` line 5625 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-NnuUKNfEzMGBPIAeTDT:-NoOF2Td6Jd6Qp09PDXj:b499q7j) ([raw file](https://github.com/cockroachdb/cockroach/blob/8607ad4c151bd720f808e5b5cb63ae0b82d8441e/pkg/kv/kvclient/kvcoord/dist_sender_test.go#L5625)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: you can use `kvpb.NewGet()` here instead.
</blockquote></details>

Changed
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_test.go` line 5640 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-NnuVh4y7dNNH5xSvKBt:-NoO-5ru1tnOxfHAtE9Y:b2kotpg) ([raw file](https://github.com/cockroachdb/cockroach/blob/8607ad4c151bd720f808e5b5cb63ae0b82d8441e/pkg/kv/kvclient/kvcoord/dist_sender_test.go#L5640)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I think this test can be simplified to only use a single Send.
</blockquote></details>

I left it in there just to make sure that are cache is actually updated. Let me think about how to do this better.

**Review by arulajmani — COMMENTED** [review]
> Reviewed 2 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107657)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2414 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-NnuRHyjDzHFXTcLmFG5:-Np_eeD7512F2--WB5Nf:b-4rknpk) ([raw file](https://github.com/cockroachdb/cockroach/blob/8607ad4c151bd720f808e5b5cb63ae0b82d8441e/pkg/kv/kvclient/kvcoord/dist_sender.go#L2414)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

Makes sense I inverted the condition. This was easier to do after rebasing on the other branch.
</blockquote></details>

Did you mean to invert this? (it's fine either way, just asking because the comment doesn't line up with the code change).
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 590 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-NnuR3pFDGxsL5_hDqb_:-Np_fGZeBMTEsMMGxU69:b-3ypprn) ([raw file](https://github.com/cockroachdb/cockroach/blob/8607ad4c151bd720f808e5b5cb63ae0b82d8441e/pkg/kv/kvclient/rangecache/range_cache.go#L590)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

It doesn't work without this check. This is because the check below is after we have already called `evictDescLocked` so it is too late.
</blockquote></details>

I may be missing something, but your comment contradicts the code change here.
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 624 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-NnuPyGwD83JyG_imtBd:-Np_dgxq6Bl1Wb7bbCGx:b1z46nu) ([raw file](https://github.com/cockroachdb/cockroach/blob/8607ad4c151bd720f808e5b5cb63ae0b82d8441e/pkg/kv/kvclient/rangecache/range_cache.go#L624)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

I'm not positive if this assertion is true or not. I changed to true/false as we can treat non-overlapping as false.
</blockquote></details>

Fair enough. I don't think it can be called with non-intersecting descriptors, but it's hard to say for certain.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_test.go` line 5570 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-NnuTcQH2mLrqPzVw_pJ:-Np_uckN0IfOCkiEzMyp:b2y83qa) ([raw file](https://github.com/cockroachdb/cockroach/blob/8607ad4c151bd720f808e5b5cb63ae0b82d8441e/pkg/kv/kvclient/kvcoord/dist_sender_test.go#L5570)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

Yeah - this is a little complex. I'm going to try and simplify this a little bit....
</blockquote></details>

Did we want to simplify this test before merging? One way to simplify this test would be to directly invoke `sendToReplicas` instead of going through `DistSender.Send`. That way,  we won't be auto-retrying in the outer loop -- instead, we'd be able to see if `sendToReplicas` is correctly breaking out or not.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_test.go` line 5647 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-Np_w1gZAtomiYYVTHOE:-Np_w1gZAtomiYYVTHOF:b-tcdga1) ([raw file](https://github.com/cockroachdb/cockroach/blob/8607ad4c151bd720f808e5b5cb63ae0b82d8441e/pkg/kv/kvclient/kvcoord/dist_sender_test.go#L5647)):*
> ```Go
> 				}
> 			}
> 			require.Equal(t, tc.expectedIterations, count)
> ```

Any reason to get rid of this?

**Review by andrewbaptist — COMMENTED** [review]
Thanks again, one more pass.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-:-NpfLSzA7LDUu_g25FVF:b-vg2jua)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2414 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-NnuRHyjDzHFXTcLmFG5:-NpfLVI_7KI7sikDokMc:b-bpsogm) ([raw file](https://github.com/cockroachdb/cockroach/blob/8607ad4c151bd720f808e5b5cb63ae0b82d8441e/pkg/kv/kvclient/kvcoord/dist_sender.go#L2414)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Did you mean to invert this? (it's fine either way, just asking because the comment doesn't line up with the code change).
</blockquote></details>

No - I had tried it first, but it got complicated with the logic below. So I ended up revertings. Sorry!
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 590 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-NnuR3pFDGxsL5_hDqb_:-NpfLvE4-ct5NHP4o7B4:b-cema39) ([raw file](https://github.com/cockroachdb/cockroach/blob/8607ad4c151bd720f808e5b5cb63ae0b82d8441e/pkg/kv/kvclient/rangecache/range_cache.go#L590)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I may be missing something, but your comment contradicts the code change here.
</blockquote></details>

Yep - I was wrong about this also. This wasn't needed as you mentioned in your first comment.
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 624 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-NnuPyGwD83JyG_imtBd:-NpfMCoJ7CWZKgtTI3Tp:b23hpoe) ([raw file](https://github.com/cockroachdb/cockroach/blob/8607ad4c151bd720f808e5b5cb63ae0b82d8441e/pkg/kv/kvclient/rangecache/range_cache.go#L624)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Fair enough. I don't think it can be called with non-intersecting descriptors, but it's hard to say for certain.
</blockquote></details>

Yeah, I wasn't clear what contract the server had about returning ranges and didn't want to codify this. It's probably safe to just leave without an assertion. 
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_test.go` line 5647 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-Np_w1gZAtomiYYVTHOE:-NpfhRy3A6gK8HZGjgqN:b-k94uzw) ([raw file](https://github.com/cockroachdb/cockroach/blob/8607ad4c151bd720f808e5b5cb63ae0b82d8441e/pkg/kv/kvclient/kvcoord/dist_sender_test.go#L5647)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Any reason to get rid of this?
</blockquote></details>

I don't actually remember other than it felt like it was too dependant on the actual code structure than testing anything worthwhile. Small refactors on the looping code change this behavior, and it wasn't really testing what it should have. I'm going to refactor this test now to only call `sendToReplicas` since all the code changes are there now and it will be a cleaner test.

**Comment by andrewbaptist** [comment]
@arulajmani one more pass at this please. I made some changes to move the overlapping function into dist_sender since it doesn't need anything from the range cache, and I cleaned up the test to really focus on `sendToReplicas`. The test is maybe still a little hard to follow, but hopefully makes sense now. Let me know if you want to chat today or tomorrow on it.

**Review by arulajmani — COMMENTED** [review]
> The test is maybe still a little hard to follow, but hopefully makes sense now.

I do think the test can be simplified even more. In particular, instead of handling 2nd store, we could always have it return success. 
This allows us to test your change in a targeted way -- we're simply asserting whether we bailed early with a `RangeKeyMismatchError` or if we tried the 2nd store. It's only two cases, which are easier to follow. 

> Reviewed 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-:-NqUrbzgDFqteRZUDWgB:b-w4vpzp)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2675 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-NqUmZJC-kF1mqijg46X:-NqUmZJC-kF1mqijg46Y:b-j0zv66) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4ed44d049bea0c196aa998c3a092b67c5df93d5/pkg/kv/kvclient/kvcoord/dist_sender.go#L2675)):*
> ```Go
> 
> // areOverlappingDescriptorsNewer compares the descriptor in this EvictionToken
> // to the descriptors passed in and returns a value whether the new descriptors
> ```

nit: "returns whether"
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2676 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-NqUnPJS7vULGREFXAYO:-NqUnPJS7vULGREFXAYP:bq1jb1e) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4ed44d049bea0c196aa998c3a092b67c5df93d5/pkg/kv/kvclient/kvcoord/dist_sender.go#L2676)):*
> ```Go
> // areOverlappingDescriptorsNewer compares the descriptor in this EvictionToken
> // to the descriptors passed in and returns a value whether the new descriptors
> // are replacements for the existing descriptor. Specifically a return of
> ```

Now that we've changed this method to return a boolean value, the comment about return values is stale.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2679 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-NqUmhRNAKdP86QxeriS:-NqUmhRNAKdP86QxeriT:bpc7hh6) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4ed44d049bea0c196aa998c3a092b67c5df93d5/pkg/kv/kvclient/kvcoord/dist_sender.go#L2679)):*
> ```Go
> // negative number means that the new descriptors are not a replacement, 0 means
> // the descriptors are equivalent and a positive number means that the new
> // descriptors do overlap and replace the existing descriptor. This comparison
> ```

nit: we shouldn't be so prescriptive about the usage of this method, lest it evolve over time. I'd recommend either moving the comment related to the usage to the caller, or wording it as "One place this comparison is useful..." and making it a different paragraph.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_test.go` line 5791 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-NqUom0O791_LAnb0XSb:-NqUom0O791_LAnb0XSc:b-19lj93) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4ed44d049bea0c196aa998c3a092b67c5df93d5/pkg/kv/kvclient/kvcoord/dist_sender_test.go#L5791)):*
> ```Go
> 					require.NoError(t, br.Error.GoError())
> 				} else {
> 					require.ErrorContains(t, br.Error.GoError(), "outside of bounds of range")
> ```

 nit: let's use:

```go
mismatchErr := &kvpb.RangeKeyMismatchError{}  
require.ErrorAs(t, br.Error.GoError(), &mismatchErr)
```

**Review by andrewbaptist — COMMENTED** [review]
I agree, and simplified the test. This is much more readable now and still provides the same coverage.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-:-Nvlbf-rEJTI6rT9uuNG:b-nmjcrc)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2675 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-NqUmZJC-kF1mqijg46X:-Nvli-pi7nivgfbLt7iE:bd1tfzi) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4ed44d049bea0c196aa998c3a092b67c5df93d5/pkg/kv/kvclient/kvcoord/dist_sender.go#L2675)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: "returns whether"
</blockquote></details>

I cleaned up this comment a lot.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2676 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-NqUnPJS7vULGREFXAYO:-Nvli1ze3oBWOLL_21Jt:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4ed44d049bea0c196aa998c3a092b67c5df93d5/pkg/kv/kvclient/kvcoord/dist_sender.go#L2676)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Now that we've changed this method to return a boolean value, the comment about return values is stale.
</blockquote></details>

Fixed.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2679 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-NqUmhRNAKdP86QxeriS:-Nvli3BS8oFaraUrQRa3:b-ni82t1) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4ed44d049bea0c196aa998c3a092b67c5df93d5/pkg/kv/kvclient/kvcoord/dist_sender.go#L2679)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: we shouldn't be so prescriptive about the usage of this method, lest it evolve over time. I'd recommend either moving the comment related to the usage to the caller, or wording it as "One place this comparison is useful..." and making it a different paragraph.
</blockquote></details>

Agreed, changed.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_test.go` line 5570 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107657#-NnuTcQH2mLrqPzVw_pJ:-Nvli6K-6X2-ENp8xfO5:brsjccd) ([raw file](https://github.com/cockroachdb/cockroach/blob/8607ad4c151bd720f808e5b5cb63ae0b82d8441e/pkg/kv/kvclient/kvcoord/dist_sender_test.go#L5570)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Did we want to simplify this test before merging? One way to simplify this test would be to directly invoke `sendToReplicas` instead of going through `DistSender.Send`. That way,  we won't be auto-retrying in the outer loop -- instead, we'd be able to see if `sendToReplicas` is correctly breaking out or not.
</blockquote></details>

Changed to just call `sendToReplicas` and simplify the additional cases.
